File: 501.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def findMode(self, root: TreeNode) -> List[int]:
        orderedList = []
        def traversal(root):#中序遍历得到有序数组
            if not root:
                return
            traversal(root.left)
            orderedList.append(root.val)
            traversal(root.right)
        traversal(root)

        count,maxCount = 0,0
        result = []
        for i in range(0,len(orderedList)):
            if i==0:
                count = 1
            elif orderedList[i]==orderedList[i-1]:
                count += 1
            else:#遇到新元素重新计数
                count = 1
            if count>maxCount:
                maxCount = count
                result.clear()
                result.append(orderedList[i])
            elif count==maxCount:
                result.append(orderedList[i])

        return result

File: 501.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    int maxCount,count;
    TreeNode* pre;
    vector<int> result;
    void traversal(TreeNode* cur){
        if(!cur) return;
        traversal(cur->left);
        if(!pre){
            count = 1;
        }
        else if(cur->val == pre->val){
            count++;
        }
        else{
            count = 1;
        }

        if(count>maxCount){
            maxCount = count;
            result.clear();//清除之前所有内容，因为不再是众数
            result.push_back(cur->val);
        }
        else if(count==maxCount){
            result.push_back(cur->val);
        }
        pre = cur;
        traversal(cur->right);
        return;
    }
public:
    vector<int> findMode(TreeNode* root) {
        maxCount = 0;
        count = 0;
        pre = nullptr;
        traversal(root);
        return result;
    }
};